<img src="ST101_a.jpeg",width=400>
# Udacity Free Trial Screener

## Experiment Overview:

## group A:
control group - when a student clicks on the old "start free trial" button, they're asked to enter their credit card info

## group B:
experimental group - new "start free trial" button asks how much time they had available to devote to the course.  If they say 5 or more, they're taken to the ususal credit card page.  Else, a message appears saying that courses take a greater time commitment and suggests staying on on free access of course materials.

<img src="Final Project_ Experiment Screenshot.png",width=600,height=600>

## hypothesis:
-  Discouraging students who have less than 5 hours/week would reduce churn of trial users without affecting completion rate.
- Business impact:
    * Improved overall student experience
    * Coaches don't waste time on deadbeats

# Metric Choice

<img src="20170207_210734.jpeg",width=600,height=600>

The screening message does not affect these metrics:  number of cookies, number of user-ids, number of clicks, click-through-probability.

Gross conversion deviates from our hypothesis since we're questioning churn not enrollment.
If we were interested in business profit, we would use common sense and scrap this experiment.

Retention doesn't capture the students who got weeeded out by the screener, which is presumably, direcly influenced by the new change.  By definition of a screener, the number of users who complete checkout in the experiment will be less; however, we can't be sure that this new design would dissuade anyone.  I choose not to use retention because # of user-ids to complete checkout is not controlled between the two groups.

- Hence, our primary evaluation metric is Net Conversion (practical signifiance boundary: dmin = .0075)

$$\frac{\text{# of user-ids to remain enrolled past trial period}}{\text{# of unique cookies to click "start free trial"}} $$

Use this metric since we're interested in seeing whether our message keeps the slackers from flaking during the trial period.

We're looking for numbers who stay past trial, which is accounted for by net conversion.  We choose this single metric because # of ! cookies to click "start free trial" is "controlled" between the two experiments.

We'll look at retention in case there was unexpected behavior, e.g. the weeder encourages more people to start a trial.

If we had time, we would want to collect an empirical estimate of the variability becasue we don't expect the analytic estimates to be accurate.

Let's estimate standard deviation of net conversion analytically, for the control group and the experimental group.

In [55]:
import pandas as pd

In [56]:
# rough estimates of the baseline values
pd.read_csv('Week 4 AB Testing Data/baseline.csv')

Unique cookies to view page per day:        40000
0  Unique cookies to click "Start free trial" per...  3200.000000
1                               Enrollments per day:   660.000000
2   Click-through-probability on "Start free trial":     0.080000
3             Probability of enrolling, given click:     0.206250
4              Probability of payment, given enroll:     0.530000
5                Probability of payment, given click     0.109313

# Control group

In [68]:
df_control = pd.read_csv('Week 4 AB Testing Data/control.csv')
df_control['Net_Conversion'] = df_control.Payments/df_control.Clicks
df_control.head()

# consider only the dates that have Payments info
df_control = df_control.dropna()
df_control

Date  Pageviews  Clicks  Enrollments  Payments  Net_Conversion
0   Sat, Oct 11       7723     687        134.0      70.0        0.101892
1   Sun, Oct 12       9102     779        147.0      70.0        0.089859
2   Mon, Oct 13      10511     909        167.0      95.0        0.104510
3   Tue, Oct 14       9871     836        156.0     105.0        0.125598
4   Wed, Oct 15      10014     837        163.0      64.0        0.076464
5   Thu, Oct 16       9670     823        138.0      82.0        0.099635
6   Fri, Oct 17       9008     748        146.0      76.0        0.101604
7   Sat, Oct 18       7434     632        110.0      70.0        0.110759
8   Sun, Oct 19       8459     691        131.0      60.0        0.086831
9   Mon, Oct 20      10667     861        165.0      97.0        0.112660
10  Tue, Oct 21      10660     867        196.0     105.0        0.121107
11  Wed, Oct 22       9947     838        162.0      92.0        0.109785
12  Thu, Oct 23       8324     665        127.0      56.0        0.084211
13  Fri, Oct 24       9434     673        220.0     122.0        0.181278
14  Sat, Oct 25       8687     691        176.0     128.0        0.185239
15  Sun, Oct 26       8896     708        161.0     104.0        0.146893
16  Mon, Oct 27       9535     759        233.0     124.0        0.163373
17  Tue, Oct 28       9363     736        154.0      91.0        0.123641
18  Wed, Oct 29       9327     739        196.0      86.0        0.116373
19  Thu, Oct 30       9345     734        167.0      75.0        0.102180
20  Fri, Oct 31       8890     706        174.0     101.0        0.143059
21   Sat, Nov 1       8460     681        156.0      93.0        0.136564
22   Sun, Nov 2       8836     693        206.0      67.0        0.096681

In [70]:
import numpy as np

# standard deviation of Net Conversion for control group
np.std(df_control.Net_Conversion)  # .0287

0.02875879092813514

In [80]:
# aggregate control_net_conversion

num_past_trial = sum(df_control.Payments)
num_clicks = sum(df_control.Clicks)
control_net_conversion = num_past_trial/num_clicks
control_net_conversion

0.11756201931417337

# Experimental group

In [71]:
df_experimental = pd.read_csv('Week 4 AB Testing Data/experiment.csv')
df_experimental['Net_Conversion'] = df_experimental.Payments/df_experimental.Clicks

# consider only the dates that have Payments info
df_experimental = df_experimental.dropna()
df_experimental

Date  Pageviews  Clicks  Enrollments  Payments  Net_Conversion
0   Sat, Oct 11       7716     686        105.0      34.0        0.049563
1   Sun, Oct 12       9288     785        116.0      91.0        0.115924
2   Mon, Oct 13      10480     884        145.0      79.0        0.089367
3   Tue, Oct 14       9867     827        138.0      92.0        0.111245
4   Wed, Oct 15       9793     832        140.0      94.0        0.112981
5   Thu, Oct 16       9500     788        129.0      61.0        0.077411
6   Fri, Oct 17       9088     780        127.0      44.0        0.056410
7   Sat, Oct 18       7664     652         94.0      62.0        0.095092
8   Sun, Oct 19       8434     697        120.0      77.0        0.110473
9   Mon, Oct 20      10496     860        153.0      98.0        0.113953
10  Tue, Oct 21      10551     864        143.0      71.0        0.082176
11  Wed, Oct 22       9737     801        128.0      70.0        0.087391
12  Thu, Oct 23       8176     642        122.0      68.0        0.105919
13  Fri, Oct 24       9402     697        194.0      94.0        0.134864
14  Sat, Oct 25       8669     669        127.0      81.0        0.121076
15  Sun, Oct 26       8881     693        153.0     101.0        0.145743
16  Mon, Oct 27       9655     771        213.0     119.0        0.154345
17  Tue, Oct 28       9396     736        162.0     120.0        0.163043
18  Wed, Oct 29       9262     727        201.0      96.0        0.132050
19  Thu, Oct 30       9308     728        207.0      67.0        0.092033
20  Fri, Oct 31       8715     722        182.0     123.0        0.170360
21   Sat, Nov 1       8448     695        142.0     100.0        0.143885
22   Sun, Nov 2       8836     724        182.0     103.0        0.142265

In [74]:
# standard deviation of Net Conversion for experimental group
np.std(df_experimental.Net_Conversion), np.var(df_experimental.Net_Conversion)

(0.031468720412915, 0.000990280364426213)

Control group net conversion standard deviation: .028

Experimental group net conversion standard deviation: .031

In [64]:
# aggregate experimental_net_conversion

num_past_trial = sum(df_experimental.Payments)
num_clicks = sum(df_experimental.Clicks)
experimental_net_conversion = num_past_trial/num_clicks
experimental_net_conversion

0.1126882966396292

In [65]:
d = abs(control_net_conversion - experimental_net_conversion)
d

0.0048737226745441675

## Sizing, choosing number of samples given power

Using the analytic estimates of variance, how many pageviews total (across both groups) would you need to collect to adequately power the experiment? Use an alpha of 0.05 and a beta of 0.2. Make sure you have enough power for each metric.

In [82]:
import statsmodels.stats.api as sms

p1 = control_net_conversion
p2 = experimental_net_conversion
es = sms.proportion_effectsize(p1, p2)
sms.NormalIndPower().solve_power(es, power=0.8, alpha=0.05, ratio=1)

67316.54367681299

So we would need a sample size of n = 67,317 clicks to obtain the desired power of 80%.